README.md

```diff
@@ -178,17 +193,62 @@ java -jar java-function-invoker-1.0.0-alpha-2-rc4.jar \
 ```
 
 
+## Running a function with Gradle
+
+From Gradle, similarily to running functions with the Functions Framework jar,
+we can invoke the `Invoker` class with a `JavaExec` task.
+
+### Configuration in `build.gradle`
+
+```groovy
+configurations {
+    invoker
+}
+
+dependencies {
+    implementation 'com.google.cloud.functions:functions-framework-api:1.0.0-alpha-2-rc3'
+    invoker 'com.google.cloud.functions.invoker:java-function-invoker:1.0.0-alpha-2-rc4'
+}
+
+tasks.register("runFunction", JavaExec) {
+    main = 'com.google.cloud.functions.invoker.runner.Invoker'
+    classpath(configurations.invoker)
+    inputs.files(configurations.runtimeClasspath, sourceSets.main.output)
+    args(
+            '--target', project.findProperty('runFunction.target'),
+            '--port', project.findProperty('runFunction.port') ?: 8080
+    )
+    doFirst {
+        args('--classpath', files(configurations.runtimeClasspath, sourceSets.main.output).asPath)
```

Comment:
I get the impression that this means we will effectively do this:

```shell
java -classpath $CP com.google.cloud.functions.invoker.runner.Invoker --classpath $CP --target ... --port ...
```

where `$CP` includes the `java-function-invoker` jar. Is that right? It's probably good enough for now, but the point of the `--classpath` option is that the user's function code should not be able to see `java-function-invoker`, which avoids problems if the user code uses a library that `java-function-invoker` also uses but in a different version. Is there a way to achieve this?

Reply:
I'm a little bit puzzled by your command-line. Why do you have a `-classpath` and `--classpath` flags? I feel like I'm missing something.

For simplicity sake for the user, I was indeed setting the classpath to be the user's own sources, as well as the dependencies defined in the `build.gradle` file.

Gradle offers two kind of scopes here, for deps, `implementation` and `api`:
https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_separation

The `api` declaration leaks the dependencies' own dependences via transitivity, whereas `implementation` won't leak them. Here's the relevant snippet in Gradle's documentation:

> Dependencies appearing in the api configurations will be transitively exposed to 
> consumers of the library, and as such will appear on the compile classpath of consumers. 
> Dependencies found in the implementation configuration will, on the other hand, not be 
> exposed to consumers, and therefore not leak into the consumers' compile classpath. 

So I think you indeed want `implementation` here, am I right?

But I don't let the user define the whole classpath himself in this Gradle build task, as I found this was a bit painful to do. So the above seemed to me like a good default. Othewise, it's still possible to update that task in one's own `build.gradle`.

Reply:
The `-classpath` flag is interpreted by the `java` command itself and sets the class path for the system class loader. That's what loads the `com.google.cloud.functions.invoker.runner.Invoker` class. Then the `--classpath` flag is interpreted by that `Invoker` and specifies the class path that it will use for loading the user's function. The class loader it uses to do that explicitly does not have anything from the system class path, with the single exception of the Functions Framework API, the exact package `com.google.cloud.functions`. This is the environment that the function will run in if deployed to GCF, and it's also what the Maven plugin establishes, so if there's a straightforward way to achieve the same thing with Gradle that would be good. It sounds like you're saying there isn't, though, so maybe we should just add a note that describes the non-straightforward way?

One thing too is that `java-function-invoker-VVV.jar` is self-contained. If you use `java -classpath java-function-invoker-VVV.jar com.google.cloud.functions.invoker.runner.Invoker ...` or `java -jar java-function-invoker-VVV.jar ...` that will work. Not sure if that helps.

Reply:
After discussion, it does appear that we have the desired property here, which is that the function code cannot see the classes from the jar declared as `invoker`. So I think we are good to go here.